src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -491,28 +491,73 @@ private static MethodHandle makeSimpleConcat(MethodType mt, String[] constants)
         if (paramCount == 0) {
             return MethodHandles.insertArguments(newStringifier(), 0, suffix == null ? "" : suffix);
         }
+
+        String prefix = constants[0];
+        MethodHandle mh;
+        var paramType0 = mt.parameterType(0);
         if (paramCount == 1) {
-            String prefix = constants[0];
             // Empty constants will be
             if (prefix.isEmpty()) {
                 if (suffix.isEmpty()) {
-                    return unaryConcat(mt.parameterType(0));
-                } else if (!mt.hasPrimitives()) {
+                    return unaryConcat(paramType0);
+                }
+
+                if (!mt.hasPrimitives()) {
                     return MethodHandles.insertArguments(simpleConcat(), 1, suffix);
-                } // else fall-through
-            } else if (suffix.isEmpty() && !mt.hasPrimitives()) {
-                // Non-primitive argument
-                return MethodHandles.insertArguments(simpleConcat(), 0, prefix);
-            } // fall-through if there's both a prefix and suffix
-        } else if (paramCount == 2 && !mt.hasPrimitives() && suffix.isEmpty()
-                && constants[0].isEmpty() && constants[1].isEmpty()) {
+                }
+            } else if (suffix.isEmpty()) {
+                if (!mt.hasPrimitives()) {
+                    // Non-primitive argument
+                    return MethodHandles.insertArguments(simpleConcat(), 0, prefix);
+                }
+            }
+
+            mh = simpleConcat3(paramType0);
+            mh = MethodHandles.insertArguments(mh, 2, suffix);
+            return MethodHandles.insertArguments(mh, 0, prefix);
+        } else if (paramCount == 2 && constants[1] == null) {
```

Comment:
All `constants` are non-null (see `parseRecipe`)
```suggestion
        } else if (paramCount == 2 && constants[1].isEmpty()) {
```